File: Strings/First_Non_Repeating_Character/FirstNonRepeatingCharacter.java
import java.util.*;

public class FirstNonRepeatingCharacter {

    public static void main(String... args) {

        System.out.println("firstNonRepeatingChar: " + firstNonRepeatingChar("abcdefacbdfh").equals("e"));
        System.out.println("firstNonRepeatingChar: " + firstNonRepeatingChar("abcd").equals("a"));
        System.out.println("firstNonRepeatingChar: " + firstNonRepeatingChar("abcda").equals("b"));

    }

    public static String firstNonRepeatingChar(String s) {
        Map<Character, Integer> charCount = new LinkedHashMap<>();
        for(int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if(!charCount.containsKey(c)) {
                charCount.put(c, 0);
            }
            charCount.put(c, charCount.get(c) + 1);
        }
        for(Map.Entry<Character, Integer> entry: charCount.entrySet()) {
            if(entry.getValue() == 1) {
                return entry.getKey().toString();
            }
        }
        return "";
    }

}

File: Strings/First_Non_Repeating_Character/first_non_repeating_character.go
package main

import (
	"fmt"
)

func firstNonRepeatedCharacter(str string) string {
	charCount := make(map[rune]int)
	for _, c := range str {
		charCount[c]++
	}
	for _, c := range str {
		if charCount[c] == 1 {
			return string(c)
		}
	}
	return ""
}

func main() {
	fmt.Println(firstNonRepeatedCharacter("abcdabcdef") == "e")
	fmt.Println(firstNonRepeatedCharacter("abcdacdef") == "b")
	fmt.Println(firstNonRepeatedCharacter("abcda") == "b")
	fmt.Println(firstNonRepeatedCharacter("abcd") == "a")
}
